src/core/util/byte_stream.cc

```diff
@@ -158,5 +158,19 @@ const std::string GetFormattedHex(const std::uint8_t* data, std::size_t size) {
   return hex.str();
 }
 
+std::unique_ptr<std::vector<std::uint8_t>> AddressToByteVector(
```

Comment:
I like this revision even though I had other thoughts in mind. Assuming we trust RVO, we may have also considered:

```cpp
std::vector<std::uint8_t> AddressToByteVector(
    const boost::asio::ip::address& address) 
{                                                   
  bool is_v4(address.is_v4());                      
  std::vector<std::uint8_t> vec(is_v4 ? 4 : 16);
  std::copy_n(
      vec.begin(),
      (is_v4 ? 4 : 16),
      (is_v4 ? address.to_v4().to_bytes().begin()
             : address.to_v6().to_bytes().begin()));
  return vec;                                       
}
```

or avoid the return:

```cpp
void AddressToByteVector(
    std::vector<std::uint8_t>& vec,
    const boost::asio::ip::address& address) 
{ 
  bool is_v4(address.is_v4());
  vec.resize((is_v4 ? 4 : 16));
  std::copy_n(
      vec.begin(), 
      (is_v4 ? 4 : 16),
      (is_v4 ? address.to_v4().to_bytes().begin()
             : address.to_v6().to_bytes().begin()));
}
```

but either of these would've required interface refactoring elsewhere (and I also prefer the move semantics of `unique_ptr`).

With that said, I think we should use our collaborated revision (https://github.com/monero-project/kovri/pull/671/commits/02ccfa87a1c996762ff752faba4794a5c3cf00b9) 👍